README.md
<a name="readme-top"></a>

<div align="center">

<h1>✂️ Cover Minifier</h1>

A CLI minifier tool is available for Extra-Networks to improve the stability of the Diffusion WebUI. <br/>It resizes all types of model covers to accelerate the rendering process for extranetwrok showcase.

<!-- SHIELD GROUP -->

[![release][release-shield]][release-url] [![releaseDate][release-date-shield]][release-date-url] [![ciTest][ci-test-shield]][ci-test-url] [![ciRelease][ci-release-shield]][ci-release-url] <br/> [![contributors][contributors-shield]][contributors-url] [![forks][forks-shield]][forks-url] [![stargazers][stargazers-shield]][stargazers-url] [![issues][issues-shield]][issues-url]

</div>

## 📦 Installation

To install `️✂️ Cover Minifier`, run the following command:

```bash
npm install -g sd-webui-cover-minifier
```

<div align="right">

[![][back-to-top]](#readme-top)

</div>

## 🤯 Usage

You can use the `cover-minifier` command to minifier all covers of Extra-Networks:

```shell
$ cd <your-stable-diffuision-webui-dir>
$ cover-minifier
```

### Options

Cover Minifier supports the following options:

```shell
--coverWidth,      -w   Width of minifiered cover, default 300
--coverHeight,     -h   Height of minifiered cover, default 400
--imgType,         -t   Type of minifiered cover [webp|png|jpg], default is webp
--ckptPath              Ckpt folder path, default is /models/Stable-diffusion
--loraPath              Lora folder path, default is /models/Lora
--hypernetworksPath     Hypernetworks folder path, default is /models/hypernetworks
--embeddingsPath        Embeddings folder path, default is /embeddings
--help,            -h   Print basic options
--version,         -v   Print cover-minifier installed version
```

<div align="right">

[![][back-to-top]](#readme-top)

</div>

## ⌨️ Local Development

You can use Gitpod for online development:

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)][gitpod-url]

Alternatively, you can clone the repository and run the following commands for local development:

```bash
$ git clone https://github.com/canisminor1990/sd-webui-cover-minifier.git
$ cd sd-webui-cover-minifier
$ npm install
$ npm start
```

<div align="right">

[![][back-to-top]](#readme-top)

</div>

## 🤝 Contributing

<!-- CONTRIBUTION GROUP -->

> 📊 Total: <kbd>**3**</kbd>

<a href="https://github.com/canisminor1990" title="canisminor1990">
  <img src="https://avatars.githubusercontent.com/u/17870709?v=4" width="50" />
</a>
<a href="https://github.com/actions-user" title="actions-user">
  <img src="https://avatars.githubusercontent.com/u/65916846?v=4" width="50" />
</a>
<a href="https://github.com/apps/dependabot" title="dependabot[bot]">
  <img src="https://avatars.githubusercontent.com/in/29110?v=4" width="50" />
</a>

<!-- CONTRIBUTION END -->

<div align="right">

[![][back-to-top]](#readme-top)

</div>

---

#### 📝 License

Copyright © 2023 [CanisMinor][profile-url]. <br /> This project is [MIT](./LICENSE) licensed.

<!-- LINK GROUP -->

[profile-url]: https://github.com/canisminor1990
[issues-url]: https://github.com/canisminor1990/sd-webui-cover-minifier/issues/new/choose
[gitpod-url]: https://gitpod.io/#https://github.com/canisminor1990/sd-webui-cover-minifier

<!-- SHIELD LINK GROUP -->

[back-to-top]: https://img.shields.io/badge/-BACK_TO_TOP-151515?style=flat-square

<!-- release -->

[release-shield]: https://img.shields.io/npm/v/sd-webui-cover-minifier?label=✂️%20NPM
[release-url]: https://www.npmjs.com/package/sd-webui-cover-minifier

<!-- releaseDate -->

[release-date-shield]: https://img.shields.io/github/release-date/canisminor1990/sd-webui-cover-minifier?style=flat
[release-date-url]: https://github.com/canisminor1990/sd-webui-cover-minifier/releases

<!-- ciTest -->

[ci-test-shield]: https://github.com/canisminor1990/sd-webui-cover-minifier/workflows/Test%20CI/badge.svg
[ci-test-url]: https://github.com/canisminor1990/sd-webui-cover-minifier/actions/workflows/test.yml

<!-- ciRelease -->

[ci-release-shield]: https://github.com/canisminor1990/sd-webui-cover-minifier/workflows/Build%20and%20Release/badge.svg
[ci-release-url]: https://github.com/canisminor1990/sd-webui-cover-minifier/actions/workflows/release.yml

<!-- contributors -->

[contributors-shield]: https://img.shields.io/github/contributors/canisminor1990/sd-webui-cover-minifier.svg?style=flat
[contributors-url]: https://github.com/canisminor1990/sd-webui-cover-minifier/graphs/contributors

<!-- forks -->

[forks-shield]: https://img.shields.io/github/forks/canisminor1990/sd-webui-cover-minifier.svg?style=flat
[forks-url]: https://github.com/canisminor1990/sd-webui-cover-minifier/network/members

<!-- stargazers -->

[stargazers-shield]: https://img.shields.io/github/stars/canisminor1990/sd-webui-cover-minifier.svg?style=flat
[stargazers-url]: https://github.com/canisminor1990/sd-webui-cover-minifier/stargazers

<!-- issues -->

[issues-shield]: https://img.shields.io/github/issues/canisminor1990/sd-webui-cover-minifier.svg?style=flat
[issues-url]: https://img.shields.io/github/issues/canisminor1990/sd-webui-cover-minifier.svg?style=flat


CHANGELOG.md
# Changelog

# 1.0.0 (2023-05-18)

### ✨ Features

- add cli tool ([461482f](https://github.com/canisminor1990/sd-webui-cover-minifier/commit/461482f))


resize-cover.js
const fs = require('fs');
const sharp = require('sharp');

const basePath = "D:\\stable-diffusion-webui"
const coverWidth = 300
const coverHeight = 440

const convertToWebp = async (path) => {
  const files = await fs.promises.readdir(path);
  for (const file of files) {
    const filePath = `${path}/${file}`;
    const stats = await fs.promises.stat(filePath);
    if (stats.isDirectory()) {
      await convertToWebp(filePath);
    } else {
      const ext = file.split('.').pop().toLowerCase();
      if (ext === 'png' || ext === 'jpg' || ext === 'jpeg') {
        const image = sharp(filePath);
        const metadata = await image.metadata();
        if (metadata.width === coverWidth && metadata.height === coverHeight && metadata.format === 'webp') {
          continue;
        }
        const webpFilePath = filePath.replace(/\.(png|jpg|jpeg)$/, '.webp');
        await image.resize(coverWidth, coverHeight).webp().toFile(webpFilePath);
        console.log(webpFilePath)
        await fs.promises.unlink(filePath);
      }
    }
  }
};

console.log("Stable-diffusion")
convertToWebp(basePath + '\\models\\Stable-diffusion');

console.log("Lora")
convertToWebp(basePath + '\\models\\Lora');

console.log("hypernetworks")
convertToWebp(basePath + '\\models\\hypernetworks');

console.log("embeddings")
convertToWebp(basePath + '\\embeddings');

src/utils/log.ts
import chalk from 'chalk';

const emoji = {
  success: String.fromCodePoint(0x2705), // ✅
  error: String.fromCodePoint(0x274c), // ❌
  warn: String.fromCodePoint(0x1f625), // 😥
  info: String.fromCodePoint(0x1f535), // 🔵
  start: String.fromCodePoint(0x1f476), // 👶
  pack: String.fromCodePoint(0x1f4e6), // 📦
  module: String.fromCodePoint(0x26aa), // ⚪
  waiting: String.fromCodePoint(0x231b), // ⌛
  boosting: String.fromCodePoint(0x1f525), // 🔥
  done: String.fromCodePoint(0x2728), // ✨
};
class Log {
  opt = {
    align: true,
    maxLength: 8,
  };
  _getTitle(emoji: string, text: string, color?: string) {
    const { maxLength } = this.opt;
    let spacing = '';
    if (this.opt.align && text.length < maxLength) {
      for (let i = 0; i < maxLength - text.length; i++) {
        spacing += ' ';
      }
    }
    const msg = color ? `${chalk[color](text)}${spacing}` : text + spacing;
    return `${emoji}  ${msg}`;
  }
  log(...msg) {
    console.log(...msg);
  }
  success(...msg) {
    this.log(this._getTitle(emoji.success, 'Success', 'green'), ...msg);
  }
  info(...msg) {
    this.log(this._getTitle(emoji.info, 'Info', 'blue'), ...msg);
  }
  warn(...msg) {
    this.log(this._getTitle(emoji.warn, 'Warn', 'yellow'), ...msg);
  }
  waitting(...msg) {
    this.log(this._getTitle(emoji.waiting, 'Wait', 'grey'), ...msg);
  }
  start(...msg) {
    this.log(this._getTitle(emoji.start, 'Done'), ...msg);
  }
  done(...msg) {
    this.log(this._getTitle(emoji.done, 'Done'), ...msg);
  }
  error(...msg) {
    this.log(this._getTitle(emoji.error, 'Error', 'red'), ...msg);
  }
  boosting(...msg) {
    this.log(this._getTitle(emoji.boosting, 'Boosting', 'magenta'), ...msg);
  }
  pack(title, ...msg) {
    this.log(this._getTitle(emoji.pack, title, 'magenta'), ...msg);
  }
  module(title, ...msg) {
    this.log(this._getTitle(emoji.pack, title, 'cyan'), ...msg);
  }
  json(obj, title) {
    this.log(
      [this._getTitle(emoji.module, title || 'Log'), JSON.stringify(obj, null, 2)].join('\n'),
    );
  }
  dye(color, ...msg) {
    return chalk[color](...msg);
  }
}
export default new Log();


src/constants/flags.ts
export interface IFlags {
  [flag: string]: {
    type: 'boolean' | 'string' | 'number';
    shortFlag?: string;
    desc?: string;
  };
}

const FLAGS: IFlags = {
  coverWidth: { type: 'number', shortFlag: 'w', desc: 'Width of minifiered cover, default 300' },
  coverHeight: { type: 'number', shortFlag: 'h', desc: 'Height of minifiered cover, default 400' },
  imgType: {
    type: 'string',
    shortFlag: 't',
    desc: 'Type of minifiered cover [webp|png|jpg], default is webp',
  },
  ckptPath: { type: 'string', desc: 'Ckpt folder path, default is /models/Stable-diffusion' },
  loraPath: { type: 'string', desc: 'Lora folder path, default is /models/Lora' },
  hypernetworksPath: {
    type: 'string',
    desc: 'Hypernetworks folder path, default is /models/hypernetworks',
  },
  embeddingsPath: { type: 'string', desc: 'Embeddings folder path, default is /embeddings' },
  help: { type: 'boolean', shortFlag: 'h', desc: 'Print basic options' },
  version: { type: 'boolean', shortFlag: 'v', desc: 'Print cover-minifier installed version' },
};

export default FLAGS;


src/commands/minifierCover.ts
import chalk from 'chalk';
import fs from 'fs';
import { resolve } from 'path';
import sharp from 'sharp';
import log from '../utils/log.js';
const convertToWebp = async ({
  path,
  coverWidth = 300,
  coverHeight = 400,
  imgType = 'webp',
}: {
  path: string;
  coverWidth: number;
  coverHeight: number;
  imgType: 'webp' | 'png' | 'jpg';
}): Promise<number> => {
  let minifierSize = 0;

  const exist = fs.existsSync(path);

  if (!exist) {
    log.error(`Can't find ${chalk.red(path)}`);
    return 0;
  }

  log.waitting(`Minifier cover in ${chalk.gray(path)}`);

  const files = await fs.promises.readdir(path);

  for (const file of files) {
    const filePath = resolve(path, file);
    const stats = await fs.promises.stat(filePath);
    if (stats.isDirectory()) {
      await convertToWebp({ path: filePath, coverWidth, coverHeight, imgType });
    } else {
      const oldeSize = stats.size;
      // @ts-ignore
      const ext = file.split('.').pop().toLowerCase();
      if (!['png', 'jpg', 'jpeg', 'webp'].includes(ext)) continue;

      const image = sharp(filePath);
      const metadata = await image.metadata();
      if (
        metadata.width === coverWidth &&
        metadata.height === coverHeight &&
        metadata.format === (imgType === 'jpg' ? 'jpeg' : imgType)
      ) {
        continue;
      }

      const newFilePath = filePath.replace(/\.(png|jpg|webp)$/, `.${imgType}`);
      const newFilePathTemp = newFilePath + '.tmp';

      switch (imgType) {
        case 'png':
          await image.resize(coverWidth, coverHeight).png().toFile(newFilePathTemp);
          break;
        case 'jpg':
          await image.resize(coverWidth, coverHeight).jpeg().toFile(newFilePathTemp);
          break;
        case 'webp':
          await image.resize(coverWidth, coverHeight).webp().toFile(newFilePathTemp);
          break;
      }

      await image.destroy();

      try {
        await fs.promises.unlink(filePath);

        await fs.promises.rename(newFilePathTemp, newFilePath);

        const newStats = await fs.promises.stat(newFilePath);

        const newSize = newStats.size;

        log.success(
          chalk.bgGreen.black(`[Minifier to ${Math.floor((newSize / oldeSize) * 100)}%]`),
          newFilePath,
        );

        minifierSize += oldeSize - newSize;
      } catch (e: any) {
        log.warn(e.message);
        await fs.promises.unlink(newFilePathTemp);
      }
    }
  }
  return minifierSize;
};

export default convertToWebp;


src/utils/createMeowSetting.ts
import cliui from 'cliui';
import { IFlags } from '../constants/flags.js';

export default (flagsConst: IFlags) => {
  // @ts-ignore
  const ui = cliui();
  const flags: any = {};
  const options = Object.entries(flagsConst)
    .map((item) => {
      if (!item[1].desc) return '';
      flags[item[0]] = {
        shortFlag: item[1].shortFlag,
        type: item[1].type,
      };
      if (!item[1].shortFlag) return `--${item[0]}\t ${item[1].desc}`;
      return `--${item[0]}, -${item[1].shortFlag}\t ${item[1].desc}`;
    })
    .filter(Boolean)
    .join('\n');

  ui.div({
    text: 'Usage',
    padding: [1, 0, 0, 0],
  });

  ui.div({
    text: '$ cover-minifier [option] [command]',
    padding: [0, 0, 0, 4],
  });

  ui.div({
    text: 'Options',
    padding: [1, 0, 0, 0],
  });

  ui.div({
    text: options,
    width: 1000,
    padding: [0, 0, 0, 4],
  });

  return {
    help: ui.toString(),
    flags,
  };
};


src/cli.ts
import chalk from 'chalk';
import meow from 'meow';
import { resolve } from 'path';
import process from 'process';
import minifierCover from './commands/minifierCover.js';
import FLAGS_CONST from './constants/flags';
import createMeowSetting from './utils/createMeowSetting';
import log from './utils/log.js';

const { help, flags } = createMeowSetting(FLAGS_CONST);

const cli = meow(help, {
  importMeta: import.meta,
  flags,
});

const run = async (flags: any) => {
  const basePath = process.cwd();
  const coverWidth = flags.coverWidth || 300;
  const coverHeight = flags.coverWidth || 440;
  const imgType = flags.imgType || 'webp';
  const ckptPath = resolve(basePath, flags.ckptPath || 'models/Stable--diffusion');
  const loraPath = resolve(basePath, flags.loraPath || 'models/Lora');
  const hypernetworksPath = resolve(basePath, flags.hypernetworksPath || 'models/hypernetworks');
  const embeddingsPath = resolve(basePath, flags.embeddingsPath || 'embeddings');

  log.info(chalk.gray('------------------------------------------------------------------'));
  log.info(chalk.gray('                    ✂️ Cover Minifier Config'));
  log.info(chalk.gray('------------------------------------------------------------------'));
  log.info('coverWidth:', chalk.blue(coverWidth));
  log.info('coverHeight:', chalk.blue(coverHeight));
  log.info('basePath:', chalk.blue(basePath));
  log.info('ckptPath:', chalk.blue(ckptPath));
  log.info('loraPath:', chalk.blue(loraPath));
  log.info('hypernetworksPath:', chalk.blue(hypernetworksPath));
  log.info('embeddingsPath:', chalk.blue(embeddingsPath));
  let minifierSize = 0;
  log.info(chalk.gray('------------------------------------------------------------------'));
  log.info(chalk.gray('                           Stable--diffusion'));
  log.info(chalk.gray('------------------------------------------------------------------'));
  minifierSize += await minifierCover({
    path: ckptPath,
    coverHeight,
    coverWidth,
    imgType,
  });
  log.info(chalk.gray('------------------------------------------------------------------'));
  log.info(chalk.gray('                           Lora'));
  log.info(chalk.gray('------------------------------------------------------------------'));
  minifierSize += await minifierCover({
    path: loraPath,
    coverHeight,
    coverWidth,
    imgType,
  });
  log.info(chalk.gray('------------------------------------------------------------------'));
  log.info(chalk.gray('                           Hypernetworks'));
  log.info(chalk.gray('------------------------------------------------------------------'));
  minifierSize += await minifierCover({
    path: hypernetworksPath,
    coverHeight,
    coverWidth,
    imgType,
  });
  log.info(chalk.gray('------------------------------------------------------------------'));
  log.info(chalk.gray('                           Embeddings'));
  log.info(chalk.gray('------------------------------------------------------------------'));
  minifierSize += await minifierCover({
    path: embeddingsPath,
    coverHeight,
    coverWidth,
    imgType,
  });
  log.info(chalk.gray('------------------------------------------------------------------'));
  log.done(`Total save`, chalk.bgGreen.black(`[${(minifierSize / 1024 / 1024).toFixed(2)}MB]`));
  log.info(chalk.gray('------------------------------------------------------------------'));
};

run(cli.flags);
